### PR TITLE
Fix clippy warnings

### PR DESCRIPTION
**Pull request by little-dude**
This fixes a bunch of clippy warnings. I grouped by warning to make it easier to review, let me know if you'd like to me split it another way.

- 51c93566d2ff78957e73b5752da705082f4a308a converts all `0 as *mut` into `ptr::null_mut()` and `0 as *const` into `ptr::null()`. I didn't change the `0 as *mut *mut T` into `ptr_null_mut()` because I was not sure is that was really equivalent.

- I'm not 100% sure about acd17f996a40470d5495c3cde55007fa5b1e5034: maybe the `&& false` was on purpose? In that case, we should probably add a comment explaining why though.

- I'm not sure about 5f0c936e2eca230d639cce720e35deaabbb40344. I mean, it looks fine to me, but maybe there is a good reason for not deriving `Clone`?

There are still a few warnings, but it compiles without warning with:
```cargo clippy -- -A many_single_char_names -A useless_let_if_seq -A cyclomatic_complexity -A needless_lifetimes -A should_implement_trait -A new_without_default_derive -A zero_ptr -A wrong_self_convention```

I'm tempted to address the `needless_lifetimes` warnings but I'm not too sure about changing the signature of unsafe function/methods in some cases.

### CODE REVIEW

**Comment by little-dude** [comment]
Ah I forgot to mention that 3a6af7f63818f07219345a7ae99a087b8ca754eb add a bunch of `is_empty()` methods to conform to https://github.com/Manishearth/rust-clippy/wiki#len_without_is_empty.

File: examples/status.rs [hunk 1]
```diff
@@ -312,8 +312,8 @@ fn print_short(repo: &Repository, statuses: git2::Statuses) {
             b = diff.new_file().path();
         }
         if let Some(diff) = entry.index_to_workdir() {
-            a = a.or(diff.old_file().path());
-            b = b.or(diff.old_file().path());
+            a.or_else(|| diff.old_file().path());
```
Comment (alexcrichton) [inline examples/status.rs]: Ah I think these lines have a different meaning now?

File: examples/status.rs [hunk 2]
```diff
@@ -338,9 +336,7 @@ fn print_short(repo: &Repository, statuses: git2::Statuses) {
 impl Args {
     fn format(&self) -> Format {
         if self.flag_short { Format::Short }
-        else if self.flag_long { Format::Long }
-        else if self.flag_porcelain { Format::Porcelain }
-        else if self.flag_z { Format::Porcelain }
+        else if self.flag_porcelain || self.flag_z { Format::Porcelain }
```
Comment (alexcrichton) [inline examples/status.rs]: Could this stay as it was? I think it was a bit clearer before

File: examples/status.rs [hunk 3]
```diff
@@ -297,9 +297,7 @@ fn print_short(repo: &Repository, statuses: git2::Statuses) {
         if let Some(status) = status {
             if status.contains(git2::SUBMODULE_STATUS_WD_MODIFIED) {
                 extra = " (new commits)";
-            } else if status.contains(git2::SUBMODULE_STATUS_WD_INDEX_MODIFIED) {
-                extra = " (modified content)";
-            } else if status.contains(git2::SUBMODULE_STATUS_WD_WD_MODIFIED) {
```
Comment (alexcrichton) [inline examples/status.rs]: Could this stay as it was before? I think it was a little clearer that way

File: examples/log.rs [hunk 4]
```diff
@@ -60,7 +60,7 @@ fn run(args: &Args) -> Result<(), Error> {
     } else {
         git2::SORT_NONE
     });
-    for commit in args.arg_commit.iter() {
+    for commit in &args.arg_commit {
```
Comment (alexcrichton) [inline examples/log.rs]: I personally prefer `for x in y.iter()` over `for x in &y`, could this commit be backed out?

File: src/diff.rs [hunk 5]
```diff
@@ -769,7 +769,6 @@ impl<'a> DiffLine<'a> {
     ///  * `B` - Line binary
     pub fn origin(&self) -> char {
         match unsafe { (*self.raw).origin as raw::git_diff_line_t } {
-            raw::GIT_DIFF_LINE_CONTEXT => ' ',
```
Comment (alexcrichton) [inline src/diff.rs]: Could this stay here? It's intended to be explicitly matched on

File: src/error.rs [hunk 6]
```diff
@@ -47,8 +47,6 @@ impl Error {
     /// Return the error code associated with this error.
     pub fn code(&self) -> ErrorCode {
         match self.raw_code() {
-            raw::GIT_OK => super::ErrorCode::GenericError,
-            raw::GIT_ERROR => super::ErrorCode::GenericError,
```
Comment (alexcrichton) [inline src/error.rs]: Could this (and other deletions from this file in this commit) stay? They're intended to be listed in the matches.

File: src/stash.rs [hunk 7]
```diff
@@ -106,7 +106,6 @@ pub extern fn stash_cb(index: size_t,
 
 fn convert_progress(progress: raw::git_stash_apply_progress_t) -> StashApplyProgress {
     match progress {
-        raw::GIT_STASH_APPLY_PROGRESS_NONE => StashApplyProgress::None,
```
Comment (alexcrichton) [inline src/stash.rs]: Could this also stay as it was before?

File: examples/clone.rs [hunk 8]
```diff
@@ -49,7 +49,7 @@ fn print(state: &mut State) {
         0
     };
     let kbytes = stats.received_bytes() / 1024;
-    if stats.received_objects() == stats.total_objects() && false {
+    if stats.received_objects() == stats.total_objects() {
```
Comment (alexcrichton) [inline examples/clone.rs]: Ah yeah I think this was just a mistake, looks good to me!

File: src/oid.rs [hunk 9]
```diff
@@ -116,10 +116,6 @@ impl Ord for Oid {
     }
 }
 
-impl Clone for Oid {
-    fn clone(&self) -> Oid { *self }
-}
```
Comment (alexcrichton) [inline src/oid.rs]: Yeah this looks ok, thanks!

File: examples/status.rs (hunk 1 above)
Comment (little-dude) [inline examples/status.rs]: oops, indeed.

File: src/error.rs (hunk 6 above)
Comment (little-dude) [inline src/error.rs]: Yeah, I'll revert this commit.

**Comment by alexcrichton** [comment]
Looks like CI is failing on Windows?

**Comment by little-dude** [comment]
Sorry, I'll try get back to this tonight or tomorrow morning. There's still one or two comments to address.

The Windows issue is strange, it works for:

```
Environment: TARGET=x86_64-pc-windows-gnu, MSYS_BITS=64
Environment: TARGET=i686-pc-windows-gnu, MSYS_BITS=32
```

But not for:

```
Environment: TARGET=x86_64-pc-windows-msvc
Environment: TARGET=i686-pc-windows-msvc
```

Googling `MSYS_BITS` leads only to `appveyor.yml` files for Rust projects on Github.

Fwiw, this was failing already for af97cb8508bc886f80292507412e7aac82f6223e so I think this is not related to this PR.

**Comment by little-dude** [comment]
Link to the failing build for af97cb8: https://ci.appveyor.com/project/alexcrichton/git2-rs/build/1.0.893

**Comment by alexcrichton** [comment]
Oh oops sorry about that! I pushed a fix to master and this otherwise looks great to me, so I'll merge. Thanks again!

**Pull request merged**